src/Surfnet/StepupSelfService/SelfServiceBundle/Controller/SamlController.php

```diff
@@ -18,29 +18,100 @@
+            $this->get('logger')->error(
+                sprintf(
+                    'Identity "%s" tried to test second factor "%s", but does not own that second factor or it is not vetted',
+                    $identity->id,
+                    $secondFactorId
+                )
+            );
+
+            throw $this->createNotFoundException();
+        }
+
+        $loaResolutionService         = $this->get('surfnet_stepup.service.loa_resolution');
+        $authenticationRequestFactory = $this->get('self_service.test_second_factor_authentication_request_factory');
+
+        $secondFactor     = $secondFactorService->findOneVetted($secondFactorId);
+        $secondFactorType = new SecondFactorType($secondFactor->type);
+
+        $authenticationRequest = $authenticationRequestFactory->createSecondFactorTestRequest(
+            $identity->nameId,
+            $loaResolutionService->getLoaByLevel($secondFactorType->getLevel())
+        );
+
+        $this->get('session')->set('second_factor_test_mode', true);
+
+        $this->get('logger')->notice(
+            sprintf(
+                'Sending authentication request with ID "%s" to the second factor test IDP',
+                $authenticationRequest->getRequestId()
+            )
+        );
+
+        return $this->get('surfnet_saml.http.redirect_binding')->createRedirectResponseFor($authenticationRequest);
+    }
+
     public function consumeAssertionAction(Request $httpRequest)
     {
-        /** @var \Surfnet\SamlBundle\Http\PostBinding $postBinding */
+        $this->get('logger')->notice('Received an authentication response for testing a second factor');
+
+        $session = $this->get('session');
+
+        if (!$session->has('second_factor_test_mode')) {
+            $this->get('logger')->error(
+                'Received an authentication response for testing a second factor, but no second factor test response was expected'
+            );
+
+            throw $this->createAccessDeniedException('Did not expect an authentication response');
+        }
+
+        $session->remove('second_factor_test_mode');
+
         $postBinding = $this->get('surfnet_saml.http.post_binding');
 
-        /** @var \SAML2_Assertion $assertion */
-        $assertion = $postBinding->processResponse(
-            $httpRequest,
-            $this->get('surfnet_saml.remote.idp'),
-            $this->get('surfnet_saml.hosted.service_provider')
-        );
+        $translator = $this->get('translator');
+
+        try {
+            $postBinding->processResponse(
+                $httpRequest,
+                $this->get('self_service.second_factor_test_idp'),
+                $this->get('surfnet_saml.hosted.service_provider')
+            );
+
+            $session->getFlashBag()->add(
+                'success',
+                $translator->trans('ss.test_second_factor.verification_successful')
```

Comment:
I noticed from the logs that this string is being translated twice: once here in the controller, and once more in the `session_flash` macro that is provided by the MopaBootstrapBundle. This triggers a warning, because the already translated string cannot be translated again.

However, when I remove the translation here, the string is no longer recognized as a translation key by the JMSTranslationBundle, so it will be removed from the result of the `extract-translations` script.

I have found a `translations.twig` file in the SelfServiceBundle which (I presume) exists to work around this problem. Is that correct? So then I could remove the translator here and add the translation keys to that file?